### ISSUE TITLE
add EverypipeTimeoutSecond to ForEach-Object-Parallel

### ISSUE THREAD

**Issue by kasini3000** [PowerShell/PowerShell#18027]
### Summary of the new feature / enhancement

# Context:
Q:how to limit time for everypipe on ForEach-Object-Parallel?
Apparently, -timeoutsecond can't do that.

It can be understood like this:
If there is no "-ThrottleLimit" parameter, -TimeoutSeconds is basically as expected.
But now there is the "-ThrottleLimit" parameter, which makes -TimeoutSeconds unable to limit the runtime of each parallel.
**“-ThrottleLimit” and “-timeoutsecond” cause subsequent pipes to time out without starting execution  **

So,we need -EverypipeTimeoutSecond.
time begin: form every parallel start

@PaulHigin

----

> -timeoutsecond can't do that.

Both in the manual and in actual tests it is demonstrated that:

```powershell
1..10 | ForEach-Object -ThrottleLimit 1 -TimeoutSeconds 2 -Parallel {
	Start-Sleep -Seconds 1
	$_
}
```

**Comment by dkaszews** [comment]
I don't quite get this. If you have 10 jobs taking 1 second each and `-ThrottleLimit` 4, so that each thread takes 2.5s on average, which do you want to limit? Single job or single thread?

**Comment by kasini3000** [comment]
I don't know if there is a one-to-one correspondence between each pipe and thread, so the name "thread" is not used here.

ForEach-Object-Parallel usage scenarios:
1Invoke-WebRequest You can't expect every pipe site to have the same return time.

When the number of pipes is large,
1 “total time” is not as easy to use as “average time”.
I don't know if there is a one-to-one correspondence between each pipe and thread, so the name "thread" is not used here.

ForEach-Object-Parallel usage scenarios:
1Invoke-WebRequest You can't expect every pipe site to have the same return time.

When the number of pipes is large,
1total time is not as easy to use as average time.
2 When a pipe hangs in front of the pipes, or the time is crowded for a long time, the subsequent pipes have not started, and “the total time” is up.

**Comment by jhoneill** [comment]
@dkaszews  I can see a wish for both. 
100 items to process, average time 1 second, 10 threads. > 10 cores to run them on. Whole process should run in 10 seconds. 
We might want to stop a single item that is still running after 10 seconds.
Or we might also want to stop the whole set if has not completed within 2 minutes 

The second seems like a terminating error "We stopped without processing everything". 
The first would be a way to fail _sooner_ but re-using the thread to process another item (maybe with a warning) strikes me as wrong - my instinct says something should process everything or throw an error, it should not silently drop items. But  I can see  someone saying "My output tells me what has been processed, the commands I'm running in parallel can't change their timeouts so let me set one in the 'command-runner'"

**Comment by kasini3000** [comment]
think:
The farmer has 10 sheep (pipes) and a trough (ThrottleLimit).
The farmer wants to average the sheep's food, not that the first sheep eat too much and the remaining 9 sheep starve to death.

**Comment by jhoneill** [comment]
> think: The farmer has 10 sheep (pipes) and a trough (ThrottleLimit). The farmer wants to average the sheep's food, not that the first sheep eat too much and the remaining 9 sheep starve to death.

The job is successfully done when no sheep are hungry. If the first sheep is removed from the trough while still hungry, do we say the job has failed ?

**Ideally** the scriptblock run for each item should limit how long it can run, and raise the right exception. 

**However** Testing this I've found that `throw` (or  `break`  or `exit`) will not stop later items from running,  So removing the "hungry sheep" is no worse than what happens already. 
For reasons I don't understand `foreach-object` doesn't allow `-parallel` and `ErrorAction/WarningAction/InformationAction` or `pipelinevariable`   -  but  setting `$ErrorActionPreference`  will cause it to stop for either  `throw` and `Write-Error`  - writing an error for "hungry sheep" would allow the other "sheep" to get their turns, but could also be transformed into a stop

**Comment by kasini3000** [comment]
The farmer said: "Don't worry about the sheep, the sick sheep will write-host or log themselves"   ;)

**Comment by kasini3000** [comment]
### news：asp.net8 pr3 add timeout for per request

https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-3/

You can set request timeouts for individual endpoints, controllers, or dynamically per request.

To apply request timeouts, first add the request timeout services:

builder.Services.AddRequestTimeouts();
You can then apply request timeouts using middleware by calling :UseRequestTimeouts()

**Comment by tolgabalci** [comment]
In my company we have hundreds of scripts that do parallel processing against many machines every day.

Prior to moving to PowerShell 7 we had to manage the jobs manual and it made the scripts about an order of magnitude more complicated than they should be.  For us, ForEach-Object is perfect and enjoy it very much.

Only caveat is randomly we will have sweeps that never return.  They go through their whole process using about 500 threads and each process finishes and shows their output, but the pipeline for ForEach-Object never returns.

I'm not clear on why this happens, but if we had a timeout per each process it would be fantastic.  It's hard to predict the full time a sweep for us will take (typically 10 to 60 minutes), however, it is very easy to determine the maximum amount a single process should take, normally no more than 10 to 30 seconds depending on the script.  Anything beyond that, it's stuck somehow and we need to stop it and free up the pool.

Some suggest ForEach-Object is not suited well for such long running processes, however I don't think that should be a scapegoat for not addressing the issues around ForEach-Object, which I think is one of the best features of PowerShell.  I understand if at a certain load it is not performant enough, however it shouldn't just crash or get stuck because you are using too many threads and memory.  If the demand is more than it can handle it should have a graceful way of reporting it.

If we can make it so each thread has a timeout limit and we can improve the termination of the processes running it would be a tremendous win for PowerShell.  I do realize processes getting stuck on IO that can't be terminated is a Windows thing, but I'm hoping the PowerShell team can help bridge the gap with the Windows team and help resolve this.  Otherwise who can?

**Comment by mklement0** [comment]
Indeed, @kasini3000, I accidentally created a duplicate of your request in:

* #20197

The initial post there contains a summary that may be helpful, as well as sample code that demonstrates the desired functionality.

**Comment by kasini3000** [comment]
share:
```powershell
#Set sets that require parallel processing
$items = 'a'..'g'

$items | ForEach-Object -ThrottleLimit 5 -Parallel {
	function f
	{
		param (
			$item
		)

		#Create a CancellationTokenSource and set 'per thread' timeout
		$timeoutInSeconds = 3 # seconds
		$cts = [System.Threading.CancellationTokenSource]::new()
		$cts.CancelAfter([TimeSpan]::FromSeconds($timeoutInSeconds))
		$token = $cts.Token

		#Simulate tasks for each project
		try
		{
			for ($i = 0; $i -lt 5; $i++)
			{
				if ($token.IsCancellationRequested)
				{
					Throw "${i}/5 runs of data source [${item}] have been canceled"
				}
				Start-Sleep -Seconds 1 # Simulate the execution time of each task
				Write-Host "Data source [${item}] has run ${i}/5 times successfully"
			}
		}
		catch
		{
			Write-Host "Data source [${item}] timeout canceled"
		}
		finally
		{
			$cts. Dispose()
		}
	}
	f  -item $_
}

```
